core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -309,6 +315,9 @@ protected static NodeShardsResult buildAllocationIdBasedNodeShardsResult(ShardRo
                 numberOfAllocationsFound++;
                 if (matchAnyShard || inSyncAllocationIds.contains(nodeShardState.allocationId())) {
                     nodeShardStates.add(nodeShardState);
+                } else if (explain) {
+                    // in explain mode, store the stale shard copies as well for reporting via the explain API
+                    nonMatchingShardStores.add(nodeShardState);
```

Comment:
this only choses those shard states which have an allocation id. Might also be interesting to have those that threw a store exception. I'm wondering if we should go even further and return all of them (also those where nothing was found).